OC-5178: Refactor S3 Parallel Operations to use latest erlware_commons

christophermaier opened the pull request:
Currently only re-tools the deletion functions (since that's all that used `erlware_commons` to begin with).  With these refactorings, it should be straightforward to refactor the file metadata retrieval functions to use `erlware_commons`, which will eliminate a lot of code.

I'll push out some further commits with that refactoring in place, but I wanted to get this code out for review as soon as possible.

Commit:
Copy ec_plists from erlware_commons

The 0.9.0 tagged release of erlware_commons depends on the master
branch of neotoma, which introduces conflicts for chef_index.  That
and other issues have prompted us to just bypass erlware_commons
altogether.

Commit:
Re-tool S3 deletion to use new erlware_commons

Uses new 0.9.0 erlware_commons to chunk and parallelize S3 delete
operations.

Also externalizes fanout and timeout parameters in app.config for more
tweakability.

seth commented on src/chef_s3_ops.erl:
```diff
@@ -33,6 +32,184 @@
+%%
+%% The spec is specifically tailored for use in this module (to make Dialyzer happy), but it
+%% is conceivable that the function could be used elsewhere as well.
+-spec fetch_and_validate_config_option(Application :: chef_objects,
```
since this is local internal helper code, any reason to keep the `Application` argument? Won't it always be `chef_objects`?

seth commented on src/chef_s3_ops.erl:
```diff
@@ -33,6 +32,184 @@
+                                                     s3_parallel_ops_timeout) ->
+                                              Value :: pos_integer().
+fetch_and_validate_config_option(Application, OptionName) ->
+    {ok, Value} = application:get_env(Application, OptionName),
```
I think this crashes here if value is unset (get_env returns `undefined`). But simple enough to just move this call into your case.

seth commented on src/chef_s3_ops.erl:
```diff
@@ -33,6 +32,184 @@
+                                              Value :: pos_integer().
+fetch_and_validate_config_option(Application, OptionName) ->
+    {ok, Value} = application:get_env(Application, OptionName),
+    case {is_integer(Value), Value > 0} of
```
Possibly more idiomatic here to use guards?

```
case Value when is_integer(Value) andalso Value > 0 ->
```

seth commented on src/chef_s3_ops.erl:
```diff
@@ -33,6 +32,184 @@
+    Key = chef_s3:make_key(OrgId, Checksum),
+
+    try mini_s3:delete_object(Bucket, Key, AwsConfig) of
+        %% Return value doesn't much matter here but it looks like, but
```
s/but/what/? Or other edit?

seth commented on src/chef_s3_ops.erl:
```diff
@@ -33,6 +32,184 @@
+            {ok, Checksum}
+    catch
+        error:{aws_error, {http_error,404,_}} ->
+            %% We got a 404.  No biggie; mark it as 'missing' and move on
```
I think we should log this. Maybe as info or warning, but would be useful to be able to track how often it happens.

seth commented on src/chef_s3_ops.erl:
```diff
@@ -33,6 +32,184 @@
+    %% Also, paranoia.
+    Token = erlang:make_ref(),
+
+    Worker = erlang:spawn(fun() ->
```
Using `proc_lib:spawn` will hook into sasl crash logging for the spawned process, let's use that.

Also wondering if we should `spawn_link`. We can discuss a bit. But I think that would make sense and would avoid spawning unlinked and unsupervised processes.

Commit:
Address code review comments

Commit:
Use ec_plists for checksum lookup, too.

This also introduces a big refactoring and reconsolidation of the
parallelization code and surrounding helpers.

Commit:
Create chef_config module

Extracts previously private functionality from chef_s3_ops

Commit:
Harmonized return types for S3 op functions

Now both checking and deleting return tuples with lists of checksums
for all of the 'ok', 'missing', 'timeout', and 'error' states.

Commit:
Tests for type-safe config retrieval

seth commented on src/chef_s3_ops.erl:
```diff
@@ -22,231 +22,294 @@
              catch
                  error:{aws_error, {http_error,404,_}} ->
+                     %% The file wasn't found.  Not logging it because this is not
+                     %% necessarily indicative of badness
```
Since the main consumer of this is via the knife cookbook upload code path, it is somewhat exceptional and I think we should log it.

seth commented on src/chef_s3_ops.erl:
```diff
@@ -22,231 +22,294 @@
+                                              Value :: pos_integer().
+fetch_and_validate_config_option(OptionName) ->
+    Application = chef_objects,
+    OptionValue = application:get_env(Application, OptionName),
```
I'm not seeing why you want `OptionValue` rather than just putting the call to `get_env` into the case. For example, where you log the error, showing the value rather than `{ok, Value}` would probably be better.

christophermaier replied:
I've fixed this in a to-be-pushed commit.

seth commented on src/chef_s3_ops.erl:
```diff
@@ -22,231 +22,294 @@
+                          MissingChecksums :: [binary()],
+                          NumTimeouts :: non_neg_integer(),
+                          NumErrors :: non_neg_integer()}.
+parallelize(OrgId, Checksums, Operation) -> %%, RequestFun) ->
```
what's the `%%, RequestFun` bit?

christophermaier replied:
That, my friend, is an artifact of several refactors :smiley:

seth commented on src/chef_s3_ops.erl:
```diff
@@ -22,231 +22,294 @@
+     NumTimeouts,
+     NumErrors}.
+
+%% @doc This bit of gymnastics is required to both simplify the implementations of
```
hmm, not sure about this approach. Might be better to export and use `?MODULE:Fun`. This will, I think, have more proper behavior under code reload situations and might give clearer stack traces if things crash inside erlware_commons. And exporting might not be so bad. From the console in debug scenario, both of these could be useful.

seth commented on src/chef_s3_ops.erl:
```diff
@@ -22,231 +22,294 @@
+                     after Timeout ->
+                             error_logger:error_msg("Operation '~p' on checksum: ~p for org ~p from bucket ~p has taken longer than ~p ms; killing spawned worker process ~p~n",
+                                                    [Operation, Checksum, OrgId, Bucket, Timeout, Worker]),
+                             erlang:exit(Worker, kill),
```
not sure about this part. We did spawn_link above so won't killing the worker kill us? I like the idea of trying to protect against a hung worker.

seth replied:
oh, so how about `erlang:unlink` before kill.

seth commented:
overall, looking good. I like the latest refactor and consolidation. Nice stuff. Comments in line. A few things that I think are worth addressing.

Commit:
Address more review comments

Commit:
More refactoring: chef_parallel

Extract timeout-aware parallelization code and make it fully generalizable for use elsewhere in Erchef.

Re-do chef_s3_ops to account for it.

Commit:
Repent for whitespace sins

Commit:
Export useful funs for debugging

Commit:
Make ec_plists dialyze with our settings

seth commented:
:+1:
